### ISSUE TITLE
MultiThreadedAugmenter: more samples generated than those of provided dataset

### ISSUE THREAD

**Issue by Seolen** [MIC-DKFZ/batchgenerators#72]
Thanks for your comprehensive augmentation implementations!
I have a question in MultiThreadedAugmenter. For my example, I fed a training dataset with 98 mini-batches to MultiThreadedAugmenter for image augmentations, and obtain an iterator that outputs 1165 mini-batches.

Q: Why it works this way? My expectation is that input 98 mini-batches and output 98 mini-batches. It is hard to define a standard epoch because traversing all output batches equals the original multiple epochs. 

----
> len(ds_train)
98
>  print(sum(1 for _ in train_loader))
1165

**Comment by FabianIsensee** [comment]
That depends on your implementation. If you intend to use DataLoader wrapped in a MultiThreadedAugmenter you need to tell DataLoader how many processes you will use (`num_threads_in_multithreaded=X`). 
Have a look at examples/multithreaded_dataloading.py
Best,
Fabian

**Comment by FabianIsensee** [comment]
Ah that example does not use the DataLoader class. The BraTS example, hoewever, does.

**Comment by Seolen** [comment]
Thanks for your replying!
Yes, I get near 12 times of training set because I set a num_threads = 12. I have wrongly understood its effect.

One more question: on the condition of not exceeding maximum memory, do larger **num_threads** and **num_cached_per_thread** mean faster data loading speed?

**Comment by FabianIsensee** [comment]
More num_threads (as long as you have enough CPU cores) will give you faster data loading speed. If you make use of probabilistic augmentations (so applying some augmentations with some probability (p_per_sample=0.3 or whatever)) then it makes sense to increase num_cached_per_thread to smoothen out the dataloading. num_cached_per_thread=3 is a good value in my experience (if you have 12 um_threads then this will save up to 36 augmented batches). If you run OOM, you can reduce that down to 1 at almost no loss in augmentation speed

**Comment by Seolen** [comment]
Thanks for your excellent interpretations again! I have got it.
BatchGenerator does provide a powerful data loading framework, and I appreciate your contribution.

**Comment by FabianIsensee** [comment]
Always happy to help
